Convert Gruntfile to javascript

fortuna opened the pull request:
Everyone is a lot more comfortable with javascript and now we have one less language to know.

We should try to get this in as soon as possible, before others modify the Gruntfile. It was a huge manual pain to get this conversion done.

This change also adds `time-grunt` and re-enables tests that were accidentally disabled in #2593.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2728)
<!-- Reviewable:end -->

fortuna commented:
@trevj, you may find it easier to review each individual commit after the automated conversion:
- [Manual fixes](https://github.com/uProxy/uproxy/pull/2728/commits/df4364165a16b5bc3f325adfdae3f431a9a40c9c)
- [More manual fixes](https://github.com/uProxy/uproxy/pull/2728/commits/536e5ab8776d3676993caf56513cbbbaf227a324)
- [Re-enable tests](https://github.com/uProxy/uproxy/pull/2728/commits/340fde065ddf15004879c47847d6908d12ba602f)

I put the automated conversion in [its own commit](https://github.com/uProxy/uproxy/pull/2728/commits/85c66be3023af89f94a404791dc8df7c78141c98) and you don't really need to review that.

fortuna commented:
FYI, this change also adds `time-grunt`:

![image](https://cloud.githubusercontent.com/assets/113565/18021516/dbfefa44-6bb5-11e6-958a-b91690fb8b41.png)

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+// Gruntfile for uProxy
```
Full stop :-) Also superfluous, though, no?

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+// Gruntfile for uProxy
+/* jshint node: true */
```
Nice.

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+module.exports = function(grunt) {
+  'use strict';
+
+  require('time-grunt')(grunt);
```
Oh, very cool.

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+  // Location of where to copy/build third_party source/libs.
+  const thirdPartyBuildPath = 'build/third_party';
+  // This is used for the copying of uproxy libraries into the target directory.
+  const localLibsDestPath = '';
```
Weird, looks like this isn't used:
https://github.com/uProxy/uproxy/issues/2729

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+      installFreedomForNodeForZork: {
+        // This allows our Docker containers, which do not have access to the
+        // git repo's "top-level" node_modules/ folder find freedom-for-node.
+
```
stray line

trevj commented on Gruntfile.js:
```diff
@@ -0,0 +1,1259 @@
+          {  // Chrome extension
+            expand: true,
+            cwd: chromeExtDevPath,
+            src: ['manifest.json', '_locales/**', 'generic_ui/scripts/copypaste.js', 'scripts/context.static.js', 'scripts/background.static.js'].concat(uiDistFiles, universalDistFiles),
```
This is very long - why not have them on separate lines? In fact, generally, should we just stick to newlines after brackets and parentheses? I'm happy for that to appear in a followup, though, given the trickiness of merging this.

trevj commented on package.json:
```diff
@@ -80,6 +81,7 @@
   },
   "scripts": {
     "build": "grunt",
+    "grunt": "grunt",
```
?

trevj commented:
OMG, thank you so much for doing this. Fewer languages FTW. Some comments but I'm very nearly happy as-is :100:

fortuna commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[Gruntfile.js, line 1 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-1:-KQLrfZsb8Dnl18HFQSf:bss4a2b) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L1)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Full stop :-) Also superfluous, though, no?
</details>
Removed.

---

*[Gruntfile.js, line 22 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-22:-KQLsey3fUXZFeEC9Rpe:b-sbbg1s) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L22)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Weird, looks like this isn't used:
> 
> https://github.com/uProxy/uproxy/issues/2729
</details>
It's used in `Rule.copyLibs()`.

---

*[Gruntfile.js, line 326 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-326:-KQLtqON8AKq-tBw98Hy:ba3dfob) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L326)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> stray line
</details>
Removed

---

*[Gruntfile.js, line 360 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-360:-KQLu4WCvSqBWdtlnGdG:b-adancr) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L360)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> This is very long - why not have them on separate lines? In fact, generally, should we just stick to newlines after brackets and parentheses? I'm happy for that to appear in a followup, though, given the trickiness of merging this.
</details>
I split the lines.
On the newline placement, I'd recommend picking whatever the Google style is (maybe the one for Javascript)

---

*[package.json, line 84 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9WoklyZ-r2-84:-KQM1Acd-pCpjnxO2e6L:bx1bi78) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/package.json#L84)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> ?
</details>
I've been using `npm run build` to run grunt without having to install it globally, but `npm run grunt` feels more natural

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2728)*
<!-- Sent from Reviewable.io -->

fortuna commented:
I've addressed the comments. Another look?

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-:-KQM3-KN_ybxQdmGIp_f:bp9b62b)*
<!-- Sent from Reviewable.io -->

trevj commented:
Still :100:.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[Gruntfile.js, line 22 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-22:-KQM8QtB-d_ds1_L9bS9:b2mebvd) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L22)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> It's used in `Rule.copyLibs()`.
</details>
But it's only set to '' so seems unnecessary. Anyway, I will investigate at some point.

---

*[Gruntfile.js, line 360 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9Wokly_-r2-360:-KQM8IXINN5M9ckkdwUO:b-n7ij08) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/Gruntfile.js#L360)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> I split the lines.
> On the newline placement, I'd recommend picking whatever the Google style is (maybe the one for Javascript)
</details>
Google JavaScript style guide sounds good to me :-)

---

*[package.json, line 84 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9WoklyZ-r2-84:-KQM7akki4r0wEm2npdL:b-9tpvxx) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/package.json#L84)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> I've been using `npm run build` to run grunt without having to install it globally, but `npm run grunt` feels more natural
</details>
Ah, it eliminates the need to install globally. Cool - and arguments are passed, so `npm run grunt zork` works as expected. Nice.

I would say in that case it's better to keep just `grunt` but I don't feel very strongly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-:-KQM8X91QdH2ayWGLLGG:b-3nevw8)*
<!-- Sent from Reviewable.io -->

fortuna commented:
I think the 100 is not enough. The presubmit check says it's missing a LGTM.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[package.json, line 84 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-KQ7nvW-H9zIQ9WoklyZ-r2-84:-KQMD6gVJ-QTHk4f5LsT:b-np73l2) ([raw file](https://github.com/uproxy/uproxy/blob/86dd5ed0b32a7d81488326ee3065d17220e92552/package.json#L84)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Ah, it eliminates the need to install globally. Cool - and arguments are passed, so `npm run grunt zork` works as expected. Nice.
> 
> I would say in that case it's better to keep just `grunt` but I don't feel very strongly.
</details>
Agreed. I've removed `npm run build`, which could be confused with `grunt build` or something else.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2728#-:-KQMCjPF-Aub8MmO6Zoy:b-fb83qx)*
<!-- Sent from Reviewable.io -->

trevj commented:
The reviewable check isn't required, only Travis - you should be able to merge!
